FOGL-4195 RPM packaging scripts revamped & its install/upgrade/remove flow fixes

ashish-jabble opened the pull request:
Signed-off-by: ashish-jabble <ashish@dianomic.com>

**Test scenarios:**

**Fresh installation**
* Installed 1.8.0 (latest develop core) - success
* Also checked purge/remove - It will keep /usr/ocal/fledge/data directory only

**Upgrade**
* Installed 1.8.1 (latest develop + bumped version and did some changes in PING response so that we know new changes are IN) over 1.8.0 - success
* Also checked purge/remove - It will keep /usr/ocal/fledge/data directory only

_Downgrade_
We have not  supported yet


**Notes**
1) You can't upgrade from release 1.8.0 to the next version (say 1.8.1). As preun & posting scriptlets are executing at the end of package. As of now I don't even have the around the patch/workaround as per RPM they were storing scriptlets into SHA's and will NOT be able to amend the scripts
Though we can see the content of scripts via `$ rpm -qa --scripts fledge` but there is NO physical location exists as such.
As we did patch placed in our debian /var/lib/dpkg/info/fledge*. but in RPM I couldn't find any related stuff.

2) You may see some warning in remove package. At the moment ignore these as it does not impact. Still investigating if I can fix around that
```
warning: file /usr/local/foglamp/data.new/extras/fogbench/fogbench_sensor_coap.template.json: remove failed: No such file or directory
warning: file /usr/local/foglamp/data.new/extras/fogbench: remove failed: No such file or directory
warning: file /usr/local/foglamp/data.new/extras: remove failed: No such file or directory
warning: file /usr/local/foglamp/data.new/etc/kerberos/README.rst: remove failed: No such file or directory
warning: file /usr/local/foglamp/data.new/etc/kerberos: remove failed: No such file or directory
warning: file /usr/local/foglamp/data.new/etc/certs: remove failed: No such file or directory
warning: file /usr/local/foglamp/data.new/etc: remove failed: No such file or directory
warning: file /usr/local/foglamp/data.new: remove failed: No such file or directory
```

YashTatkondawar commented on packages/RPM/SPECS/fledge.spec:
```diff
@@ -452,11 +503,17 @@ remove_fledge_sudoer_file() {
-remove_unused_files
+# main
+if [ $1 == 1 ];then
+    echo "postun step: ${PKG_NAME} is getting upgraded. But nothing to upgrade."
```
Can we print more precise message here? This statement seems confusing.

ashish-jabble commented on packages/RPM/SPECS/fledge.spec:
```diff
@@ -351,9 +401,10 @@ install_pip3_packages () {
 	source scl_source enable rh-python36
 
-	pip install -Ir /usr/local/fledge/python/requirements.txt
+    # TODO: we may need with --no-cache-dir
```
This is for NOTE purpose only; as we have seen this in past if we failed to download any pip package (especially aiohttp*) and it exits from here.

ashish-jabble commented on packages/RPM/SPECS/fledge.spec:
```diff
@@ -86,29 +106,59 @@ exists_schema_change_path () {
-    echo "Fledge is already installed: this is an upgrade/downgrade."
+disable_fledge_service () {
+	set +e
+	/sbin/chkconfig fledge off
```
We can but for this release its okay, I have just fixed the flow of scriptlets and other required stuff. And this is not the single reference; there are multiple references with this legacy command used in spec file (on/off)

ashish-jabble commented:
@MarkRiddoch As I have added some notes in the PR description itself. Regarding upgrade from 1.8.0 to next released version is **NOT** possible for RPM; as in our 1.8.0 **postun** scriptlet we have removed the dir content of `/usr/local/fledge` except the **data** directory. And for Debian we got a fix and had placed a workaround to make it work but couldn't find the way for RPM because that's not how this works. The previous spec file doesn't exist on the disk. The scriptlets it included are stored in the rpm database. So you can't "edit" that.

And this PR prevents the next upgrades.
Kindly advise!

**Note** - If we have any customer/user who have installed **1.8.0** RPM package, they will only need to do below manual steps

```
1) sudo apt update -  so that they get the latest packages. **And donot do sudo apt upgrade instead do below step #2**
2) sudo apt purge fledge - as it allows us to keep the data dir (i.e DB back up and other scripts related stuff)
4) sudo -H apt install -y fledge - it will install the new fledge but with the old DB which we have for the previous version
```
